src/app/forms/components/date/focus-next.directive.ts

```diff
@@ -0,0 +1,40 @@
+import { Directive, ElementRef, HostListener, Input } from '@angular/core';
+
+@Directive({
+  selector: '[appFocusNext]'
+})
+export class FocusNextDirective {
+  @Input() includeTime = false;
+
+  constructor(private el: ElementRef) {}
+
+  @HostListener('input', ['$event'])
+  onInput(e: InputEvent) {
+    const {
+      nativeElement: { id, value }
+    } = this.el;
+    const segments = id.split('-');
+    const currentSegment = segments.splice(-1)[0];
+    const name = segments.join('-');
+
+    if (value.length === 2) {
+      if ('day' === currentSegment) {
+        this.focus(`${name}-month`);
+        return;
+      } else if ('month' === currentSegment) {
+        this.focus(`${name}-year`);
+        return;
+      } else if ('hour' === currentSegment) {
+        this.focus(`${name}-minute`);
+        return;
+      }
+    } else if (value.length === 4 && this.includeTime && 'year' === currentSegment) {
+      this.focus(`${name}-hour`);
+      return;
+    }
+  }
+
+  focus(el: string) {
+    document.getElementById(el)?.focus();
+  }
```

Comment:
As requested, here is an alternative in which calculating the suffix is abstracted:
```suggestion
  @HostListener('input', ['$event'])
  onInput(e: InputEvent) {
    const segments = this.el.nativeElement.id.split('-');

    const suffix = this.getSuffix(segments.splice(-1)[0]);

    if (suffix) {
      document.getElementById(segments.join('-') + suffix)?.focus();
    }
  }

  getSuffix(currentSegment: any): string | undefined {
    const { nativeElement: value } = this.el;

    if (value.length === 2) {
      if ('day' === currentSegment) {
        return '-month';
      } else if ('month' === currentSegment) {
        return '-year';
      } else if ('hour' === currentSegment) {
        return '-minute';
      }
    } else if (value.length === 4 && this.includeTime && currentSegment === 'year') {
      return '-hour';
    }
  }
```

Reply:
Thank, good idea! I've implemented it with some small changes.